Postgres issues

mulcahyk opened the pull request:
This pull request deals primarily with two postgres issues.
First, when it comes to updating the stats in postgres the connection was being refused because of too many connections.
Second, there was a refactoring of the way that the facts are being stored in postgres.  This required a change to the way that queries that gather the stats information were written.

mulcahyk committed:
Load the bundles in parallel
Continue to walk the files the directory but then fan-out their processing to worker threads.  Each worker thread now creates its own DB connection.
Add ZipCode to the raw stats.

mulcahyk committed:
Load the bundles in parallel
Continue to walk the files the directory but then fan-out their processing to worker threads.  Each worker thread now creates its own DB connection.
Add ZipCode to the raw stats.

mulcahyk committed:
Have rawstats use patient.id as _id

mulcahyk committed:
Add Deceased and store system with condition code

mulcahyk committed:
Add County/SubCounty info to rawstats

mulcahyk committed:
Group location information together.

mulcahyk committed:
Group location information together.

mulcahyk committed:
Add disease key and use for aggregation
Get the diseasefp key from the postgres db.
Add the query to get the subcounty stats from mongo into postgres.

mulcahyk committed:
Add disease key and use for aggregation
Get the diseasefp key from the postgres db.
Add the query to get the subcounty stats from mongo into postgres.

mulcahyk committed:
Query fixup
Fixed case issue in the mongodb query.
Added all the queries for postgres.

mulcahyk committed:
Query fixup
Fixed case issue in the mongodb query.
Added all the queries for postgres.

mulcahyk committed:
Fixed postgres table updates
stats and facts tables are now being populated witht the correct data.

mulcahyk committed:
Fixed postgres table updates
stats and facts tables are now being populated witht the correct data.

mulcahyk committed:
Add scripts to clear databases for loading

mulcahyk committed:
Add scripts to clear databases for loading

mulcahyk committed:
Tweaks and touch ups from code review

mulcahyk committed:
Tweaks and touch ups from code review

mulcahyk committed:
Changes to accommodate new schema

mulcahyk committed:
Changes to accommodate new schema

mulcahyk committed:
Handle some zeros
For the fips lookup truncate " Town" from the names coming from postgres so that they match what synthea is generating.
Update the mongo queries for disease and condition stats so that only diseases and conditions with ids > 0 will be included.

mulcahyk committed:
Handle some zeros
For the fips lookup truncate " Town" from the names coming from postgres so that they match what synthea is generating.
Update the mongo queries for disease and condition stats so that only diseases and conditions with ids > 0 will be included.

mulcahyk committed:
Re-establish the postgres for updating the stats
Close the connection once the lookups are built and then open it again when doing that stats push.  The push to mongo could exceed the postgres connecton timeout.

mulcahyk committed:
Clear the fact tables before loading

mulcahyk committed:
Clear the fact tables before loading

mulcahyk committed:
Rebase fixes

mulcahyk committed:
Rebase fix

mulcahyk committed:
Merge branch 'postgres_issues' of https://github.com/synthetichealth/bulkfhirloader into postgres_issues

mulcahyk committed:
Fixes from rebasing

mulcahyk committed:
More detail for the README

gregquinn2001 commented:
Can we improve the README build section to include how you build.
Also, a section on how to add a new statistic would be very useful

cjduffett commented on database_work.go:
```diff
@@ -56,101 +94,154 @@ func CalculateFacts(mgoSession *mgo.Session, mDB string, pgConn string) {
+	for iter.Next(&result) {
+		_, err = stmt.Exec(result.ID.CsFips, result.ID.AgeRange, result.Pop, result.PopMale, result.PopFemale)
+		if err != nil {
+			fmt.Println(err)
```
There is no need to both print and log the error. `log.Fatal(err)` logs the error to stdout (using `log.Println(err)` then calls `panic(err)`. I see this in many places - could you correct these?

Alternatively you could do:

```
log.Println(err)
panic(err)
```

But `log.Fatal(err)` is more idiomatic.

cjduffett commented on database_work.go:
```diff
@@ -56,101 +94,154 @@ func CalculateFacts(mgoSession *mgo.Session, mDB string, pgConn string) {
-			on f.cousubfp = s.cs_fips
-		group by s.ct_fips, f.diseasefp`)
-	if err != nil {
+// CalculateDiseaseFact: Calculate the populations broken down by disease.  Only counting living patients.  A patient is counted only once per disease.
```
The comment does not match the function signature below. Assuming you just copy-pasted. No big deal, just fix it along with the other issues before you commit.

cjduffett commented on clear_postgres_tables.sql:
```diff
@@ -1,7 +1,6 @@
 /*
```
If everything in this file is commented out do we really need it anymore?

mulcahyk replied:
This is a practice from my previous company - scripts that do deletes like this are commented out so that you don't accidentally delete all your data.

cjduffett commented on bundle.go:
```diff
@@ -5,17 +5,107 @@ import (
+	result := []int{}
+
+	for v := range elements {
+		if encountered[elements[v].DiseaseID] == true {
```
You can make this more efficient, since duplicates will just re-assign `true` to the same key in the map.

``` go
for v := range elements {
    encountered[elements[v].DiseaseID] == true
}

// Then iterate through the map's keys
result := make([]int, len(encountered))
i := 0
for k := range encountered {
    result[i] = k
    i++
}
```

That way you're not constantly appending to result when you already know what its size will be (based on the number of keys in the `encountered` map).

cjduffett commented on bundle.go:
```diff
@@ -5,17 +5,107 @@ import (
+	result := []int{}
+
+	for v := range elements {
+		if encountered[elements[v].ConditionID] == true {
```
Same as above - you can do this more efficiently.

cjduffett commented on bundle.go:
```diff
@@ -5,17 +5,107 @@ import (
+	return result
+}
+
+func diff(a, b time.Time) (year, month, day, hour, min, sec int) {
```
If all you're trying to do is find the difference between `time.Time` objects `a` and time `b` (to determine the age of the patient), Go has built-ins that will do this for you:

``` go
// assuming b is more recent than a
delta := b.Sub(a)

// Sub() returns a Duration. The largest field in Duration is 'Hour', so you'll need to divide
// to get the total number of years
years := delta.Hours() / (24 * 365)
```

You will probably want to floor this to the nearest whole number.

Either way, it seems the only information you extract from this method is the year. If that's the case, don't bother returning the month, day, hour, etc.

cjduffett commented on database_work.go:
```diff
@@ -10,27 +11,64 @@ import (
+
+// CalculatePopulation: Calculate the basic population facts.  Only counting living patients.
+func CalculatePopulation(mgoSession *mgo.Session, mDB string, db *sql.DB) {
+	if err := db.Ping(); err != nil {
```
You're welcome to repeatedly `Ping()` the database but there is a very small chance the connection has soured since you opened it. Up to you. Just to clarify, you now only open one Postgres connection from `main()` and use it everywhere else? Assuming this has resolve the issue with opening too many connections?

cjduffett commented on README.md:
```diff
@@ -1,12 +1,44 @@
 # bulkfhirloader
```
Could you provide some additional information about the parallelization? How many workers are used? Is there a way to configure this?

cjduffett commented on README.md:
```diff
@@ -1,12 +1,44 @@
+A sample commandline would look like:
+    nohup ./bulkloader /data/fhir localhost:27017 fhir pguser:pgpwd@localhost/fhir &
+
+    - This assumes that both the Mongo and Postgres databases are called fhir and are on the localhost
```
For some reason GitHub's markdown parser is interpreting these bullet points as a code block, probably because of the indentation. List the sample command in triple quotes (```) to display it as a code block:

```
nohup ./bulkloader /data/fhir localhost:27017 fhir pguser:pgpwd@localhost/fhir &
```

Then list your bullet points using asterisks (*), without any indentation:
- point 1
- point 2
- point 3

See this guide for markdown tips: [Markdown Cheat Sheet](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)

cjduffett commented on bundle.go:
```diff
@@ -34,6 +124,9 @@ func UploadResources(resources []interface{}, mgoSession *mgo.Session, mDB strin
 			if ok {
```
`ok` means that the resource was successfully cast as a `Patient` object but it doesn't necessarily mean that all of the fields were deserialized correctly. For example, today we encountered an issue with `p.BirthDate` being `nil`. You really should error check that these fields are not nil before trying to access them. The extra error logging (in the cases where these fields _are_ nil) will also help us debug things more easily.

I know it's a PITA but it's totally worth it.

cjduffett reviewed (changes requested):
Hey Kristian, overall looks good. Made a series of comments for you to review before merging.

Commit:
Refactoring and Updates

Refactored the project's folder hierarchy to be more idiomatic. Updated
command-line flags for configurability and usability. Addressed
outstanding issues with the code and documentation addressed earlier in
this pull request. Huzzah!

cjduffett commented:
I addressed all of these issues (and Greg's request for documentation) in my latest commit. I will push this commit once Andre (or Greg?) grants me access to the bulkfhirloader repository.

Commit:
Minor fixes to README

Commit:
Added filepath.Walk error check

Added a check that filepath.Walk() didn't return an error (for example if the path to the bundles doesn't exist) before attempting to read the bundle.
